client/src/components/logViewer/logParser.ts

```diff
@@ -0,0 +1,417 @@
+}
+
+function handleProblemLocation(
+  problemMetadata: ProblemMetadata,
+  codeLogLine: string,
+  offset: LocationOffset,
+) {
+  if (!isValidProblemLocation(problemMetadata.location)) {
+    problemMetadata.location = generateLocationFromCodeLogLine(
+      codeLogLine,
+      offset,
+    );
+  }
+}
+
+function generateLocationFromIndicator(
+  logLine: string,
+  codeLogLine: string,
+  offset: LocationOffset,
+): ProblemLocation {
+  const regex = /^(?<space>\s*)(?<hyphen>[-]+)$/;
+  const found = logLine.match(regex);
+  if (found === null) {
+    return null;
+  } else {
+    const { lineNumber: codeLogLineNumber } =
+      extractCodeInfoFromCodeLogLine(codeLogLine);
+
+    const { space, hyphen } = found.groups;
+    const startColumn = space.length - offset.columnOffset;
+    const endColumn = startColumn + hyphen.length;
+
+    return {
+      lineNumber: codeLogLineNumber - offset.lineOffset,
+      startColumn,
+      endColumn,
+    };
+  }
+}
+
+function generateLocationFromCodeLogLine(
+  codeLogLine: string,
+  offset: LocationOffset,
+): ProblemLocation {
+  const { code, lineNumber: codeLogLineNumber } =
+    extractCodeInfoFromCodeLogLine(codeLogLine);
+  const startColumn =
+    getFirstCharacterColumn(code) +
+    codeLogLineNumber.toString().length -
+    offset.columnOffset;
+  const endColumn = startColumn + code.trim().length;
+  const lineNumber = codeLogLineNumber - offset.lineOffset;
+
+  return { lineNumber, startColumn, endColumn };
+}
+
+function isEmptyCodeLogLine(logLine: string): boolean {
+  return /^\d+\s*$/.test(logLine);
+}
+
+function constructProblem(problemMetadata: ProblemMetadata): Problem {
+  const {
+    type,
+    location: { lineNumber, startColumn, endColumn },
+  } = problemMetadata;
+  const message = problemMetadata.logLines
+    .map((logLine) => logLine.trim())
+    .join(" ");
+
+  return { lineNumber, startColumn, endColumn, message, type };
+}
+
+function getProblemIndexFromIndicatorLogLine(logLine: string): string | null {
+  const regex = /^\s*(?<problemIndex>\d+)$/;
+  const match = logLine.match(regex);
+  if (match === null) {
+    return null;
+  } else {
+    return match.groups.problemIndex;
+  }
+}
+
+function getProblemIndexFromProblemBeginningLogLine(logLine: string): string {
+  const regex = /^(error|warning)\s*(?<problemIndex>\d+)-\d+:\s[\w\W]*/i;
+  const match = logLine.match(regex);
+  return match.groups.problemIndex;
+}
+
+function isProblemTypeLog(logItem: LogItem): boolean {
+  return logItem.type === "error" || logItem.type === "warning";
+}
+
+function isSourceTypeLog(logItem: LogItem): boolean {
+  return logItem.type === "source";
+}
+
+// "ERROR 22-322: Syntax error, expecting one of the following: ;, CANCEL, "
+// "WARNING: Variable POP_100 not found in data set WORK.UNIVOUT.""
+function isProblemBeginningLogLine(logLine: string): boolean {
+  return /^(?<logType>error|warning)(?<errorCatagory>\s*\d+-\d+)?:\s(?<message>[\w\W]*)/i.test(
```

Comment:
there're same regexp repeated in this file, shall we make them const variables?

Reply:
?